dex/src/main/java/io/crate/data/AsyncOperationBatchIterator.java

```diff
@@ -128,10 +128,14 @@ private void concatRows(Iterator<? extends T> rows) {
     private CompletionStage<?> uncheckedLoadNextBatch() {
         CompletionStage<?> batchProcessResult = tryProcessBatchFromLoadedSource();
         if (batchProcessResult == null) {
-            if (source.allLoaded()) {
-                return processRemaining();
+            try {
+                if (source.allLoaded()) {
+                    return processRemaining();
+                }
+                return source.loadNextBatch().thenCompose(ignored -> this.uncheckedLoadNextBatch());
+            } catch (Throwable t) {
+                return CompletableFuture.failedFuture(t);
```

Comment:
Couldn't we change `uncheckedLoadNextBatch` to throw an Exception as well? It's called within `loadNextBatch`.

Reply:
Yes right, although the nested `uncheckedLoadNextBatch` call must be caught then..

Reply:
Hm k, let's keep it as is then.